Added TransIP DNS Plugin

Tim81 opened the pull request:
Added TransIP DNS Plugin

Required parameters:
-CustomerName         is equal to your loginname at TransIP
-PrivateKey          is equal to the PrivateKey you can generate at the controlpanel of TransIP
-GlobalKey         $true means that IP Whitelisting for the used PrivateKey is turned off. $false means IP Whitelisting is in use

Tim81 committed:
Add files via upload

Added TransIP DNS Plugin

Required parameters:
-CustomerName         is equal to your loginname at TransIP
-PrivateKey          is equal to the PrivateKey you can generate at the controlpanel of TransIP
-GlobalKey         $true means that IP Whitelisting for the used PrivateKey is turned off. $false means IP Whitelisting is in use

Tim81 committed:
Update TransIP.ps1 - Added comment sections

Added comment sections to each function

dapkor reviewed (commented):
Sweet, Love to see this automated !

rmbolger commented:
Addresses #622

rmbolger commented:
It looks like this plugin will currently only work on PowerShell 7+ because [ImportFromPem](https://learn.microsoft.com/en-us/dotnet/api/system.security.cryptography.rsa.importfrompem) doesn't exist in .NET Framework that old PowerShell 5.1 uses.

However, it should be possible to use some of the module's own private functions to import in a way that's compatible with both. Specifically, `Import-Pem` will load the key into a BouncyCastle AsymmetricCipherKeyPair object which can then be passed to `ConvertFrom-BCKey` to get back a .NET RSACryptoServiceProvider object that you should be able to use the same way with `SignData()` as you are now.

I'd also suggest we allow for a couple options to specify the private key. Passing in as a string is fine, but we should make it a SecureString so it doesn't get saved to disk in the clear. But some folks would also probably prefer an option that let's them just pass a file path to the key file.

Tim81 commented:
I’ll look into it, but with testing I actually used a file as source as well, but used get-content to pass it to the variable.

Tim81 committed:
Added SecureString, FilePath for PrivateKey and WinPS5.1 compatability

The plugin now supports:
- PrivateKeySecureString parameter
This way you can use a direct SecureString source, like Azure Key Vault
- PrivateKeyFilePath parameter
Use a PEM file

 PKCS#1 and PKCS#8 key formats are supported

Windows Powershell 5.1 compatability is added.
PKCS#8 key format on Windows Powershell 5.1 is only supported with .NET 4.7.2 or later (Default installed on Win10+ / Server2016+)

Tim81 commented:
@rmbolger I've added SecureString, Filepath parameters for the PrivateKey, removed the plaintext parameter. Also added Windows Powershell 5.1 compatability.

Tim81 committed:
Added support to supply own JWT token for authentication

Added support to supply own JWT token for authentication instead of generation one ourselves. When JWT token is supplied, CustomerName, PrivateKey and GlobalKey parameters are not required.

Tim81 commented:
Added support to supply the JWT access token directly.

Tim81 commented:
Hi @rmbolger when will the PR reviewed further? :-)

Tim81 commented:
I've created an account and purchased a throwaway domain to test with.
I have emailed you the credentials

Tim81 committed:
Created different parametersets for PrivateKey and changed parametername CustomerName to TIPCustomerName

Created different parametersets for PrivateKey and changed parametername CustomerName to TIPCustomerName.

As well made the DNS $RecordName and $TxtValue the first parameters, altough due to explicit position value, it shouldn't matter.

Tim81 commented:
Created different parametersets for PrivateKey and changed parametername CustomerName to TIPCustomerName, as well made the DNS $RecordName and $TxtValue the first parameters, altough due to explicit position value, it shouldn't matter.

rmbolger committed:
Refactored param names a bit.

rmbolger committed:
Rename TIPGlobalKey to TIPEnforceWhitelist and use [switch] instead of [bool]

rmbolger committed:
added debug logging for REST calls and minor code cleanup

rmbolger committed:
moved module-called functions to the top, clarified help text

rmbolger commented:
Pushed some changes that mostly revolve around parameter naming and debug logging. I need to head to bed, but I stumbled on a potential bug when I was testing adding records to the zone apex. TransIP throws an error if you try to add a record that has an existing record with the same relative name and type if the existing record's TTL doesn't match the new record's TTL.

For example, the default SPF TXT record on the domains has a TTL of 1 hour. But if the plugin tries to add another `@` record with a challenge token value and TTL of only 5 minutes, it fails. We probably need to tweak the code that checks for an existing record and grab the TTL value to re-use on the new record if one already exists with the same name+type.

Tim81 commented:
Alright, so before actually adding a record check if one exists and use it’s TTL instead of 5 minutes. Also means that we need to get the full record before removing as the API uses the TTL value there as well. Is a required value in the API to remove a TXT record, I don’t understand why if you specify the full name and it’s value, but the API demands it.

rmbolger commented:
I was actually just about to test a fix for this. The API doesn't seem to care if the TTL matches on the delete request body at least if I recall correctly from last night.

rmbolger committed:
fix bug requiring matching TTLs with existing records

rmbolger committed:
add user guide

rmbolger committed:
add plugin to supporting files

rmbolger commented:
I was just about to push a release when you submitted the PR. So likely very soon.
